crates/ruff_python_semantic/src/globals.rs

```diff
@@ -0,0 +1,95 @@
+//! When building a semantic model, we often need to know which names in a given scope are declared
+//! as `global`. This module provides data structures for storing and querying the set of `global`
+//! names in a given scope.
+
+use std::ops::Index;
+
+use ruff_text_size::TextRange;
+use rustc_hash::FxHashMap;
+use rustpython_parser::ast;
+use rustpython_parser::ast::Stmt;
+
+use ruff_index::{newtype_index, IndexVec};
+use ruff_python_ast::statement_visitor::{walk_stmt, StatementVisitor};
+
+/// Id uniquely identifying the set of global names for a given scope.
+#[newtype_index]
+pub struct GlobalsId;
+
+#[derive(Debug, Default)]
+pub(crate) struct GlobalsArena<'a>(IndexVec<GlobalsId, Globals<'a>>);
+
+impl<'a> GlobalsArena<'a> {
+    /// Inserts a new set of global names into the global names arena and returns its unique id.
+    pub(crate) fn push(&mut self, globals: Globals<'a>) -> GlobalsId {
+        self.0.push(globals)
+    }
+}
+
+impl<'a> Index<GlobalsId> for GlobalsArena<'a> {
+    type Output = Globals<'a>;
+
+    #[inline]
+    fn index(&self, index: GlobalsId) -> &Self::Output {
+        &self.0[index]
+    }
+}
+
+/// The set of global names for a given scope, represented as a map from the name of the global to
+/// the range of the declaration in the source code.
+#[derive(Debug)]
+pub struct Globals<'a>(FxHashMap<&'a str, TextRange>);
```

Comment:
I may have over-abstracted this a bit, here + with `GlobalsBuilder`? Curious to get reactions.

Reply:
I think it makes sense. I would rename it to GlobalsVisitor because it implements visitor and doesn't really provide builder methods (other than finish)

Reply:
Java Java Java

Reply:
Builder all the things